File: backend/services/post-kt-vertx-fdm/src/main/kotlin/me/jason5lee/post_kt_vertx_fdm/user_register/impl.kt
package me.jason5lee.post_kt_vertx_fdm.user_register

import com.mongodb.MongoWriteException
import kotlinx.coroutines.reactive.awaitSingle
import me.jason5lee.post_kt_vertx_fdm.common.Password
import me.jason5lee.post_kt_vertx_fdm.common.UserId
import me.jason5lee.post_kt_vertx_fdm.common.UserName
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Db
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Deps
import me.jason5lee.post_ktor_mongo_fdm.common.utils.timeFromNow
import org.bson.Document

class WorkflowImpl(private val deps: Deps) : Workflow(), ErrorsImpl {
    override suspend fun createUser(userName: me.jason5lee.post_kt_vertx_fdm.common.UserName, password: me.jason5lee.post_kt_vertx_fdm.common.Password): me.jason5lee.post_kt_vertx_fdm.common.UserId {
        val doc = Document()
        doc["name"] = userName.value
        doc["encryptedPassword"] = password.encrypt(deps.bCrypt)
        doc["creationTime"] = timeFromNow(offset = 0).utc

        val inserted = try {
            deps.mongoDb.getCollection(Db.users).insertOne(doc).awaitSingle()
        } catch (e: MongoWriteException) {
            throw userNameAlreadyExists()
        }

        return me.jason5lee.post_kt_vertx_fdm.common.UserId(Db.formatId(inserted.insertedId!!.asObjectId().value))
    }
}

File: backend/services/post-ts-koa-mongo/src/user-register/impl.ts
import { UserName, Password, UserId } from "../common";
import { Deps, now } from "../common/utils";
import { Workflow } from ".";
import * as db from "../common/utils/db";
import { Long, MongoError } from "mongodb";
import { errors } from "./api";

export class WorkflowImpl extends Workflow {
    async createUser(userName: UserName, password: Password): Promise<UserId> {
        try {
            const rec = await this.deps.mongoDb.collection(db.users).insertOne({
                name: userName,
                encryptedPassword: await password.encrypt(this.deps.encryptor),
                creationTime: Long.fromNumber(now().utc),
            } satisfies {
                name: string,
                encryptedPassword: string,
                creationTime: Long,
            });
            return db.formatId(rec.insertedId) as UserId;
        } catch (e) {
            if (e instanceof MongoError) {
                if (e.code === 11000) {
                    throw this.errors.userNameAlreadyExists();
                }
            }
            throw e;
        }
    }

    readonly errors = errors;
    constructor(private readonly deps: Deps) { super(); }
}
